File: bank/bank-python-console/bank/money.py
"""Money helpers.

**Never use float for money.** `0.1 + 0.2` is `0.30000000000000004` in Python exactly as it is in
Java, because both use binary floating point and neither can represent 0.1 exactly. Python's answer
is `decimal.Decimal` — and unlike Java's BigDecimal it keeps the normal operators, so `a + b` works
and reads like arithmetic.

The one rule that catches everyone: build a Decimal from a **string**, never from a float.
`Decimal(0.1)` is 0.1000000000000000055511151231257827 because the float was already wrong before
Decimal saw it. `Decimal("0.1")` is exactly 0.1.
"""

from __future__ import annotations

from decimal import Decimal, InvalidOperation, ROUND_HALF_UP

# Module-level constants. UPPER_CASE by convention; Python has no `final`, the naming is the
# contract. Two places is one decimal point, currency style.
CENTS = Decimal("0.01")
ZERO = Decimal("0.00")

#: The per-transaction cap, so a fat-fingered extra zero is caught rather than deposited.
MAX_TRANSACTION = Decimal("50000.00")


def money(value: str | int | Decimal) -> Decimal:
    """Build a rounded Decimal.

    `str(value)` first, so passing a float still goes through the string route rather than
    inheriting the float's error. Being forgiving here costs nothing and removes a whole class of
    bug from every caller.
    """
    return round_money(Decimal(str(value)))


def round_money(value: Decimal) -> Decimal:
    """Round to cents.

    ROUND_HALF_UP is what people expect: 2.345 becomes 2.35. Python's *default* is
    ROUND_HALF_EVEN ("banker's rounding"), which sends 2.345 to 2.34 — statistically fairer over
    many roundings, and completely baffling on a single receipt. State the rounding you want.
    """
    return value.quantize(CENTS, rounding=ROUND_HALF_UP)


def parse_money(text: str) -> Decimal:
    """Parse typed input, tolerating "$1,200.50" as well as "1200.5".

    Raises:
        ValueError: when the text is not a number at all.
    """
    cleaned = text.strip().replace("$", "").replace(",", "")
    if not cleaned:
        raise ValueError("No amount entered")
    try:
        return round_money(Decimal(cleaned))
    except InvalidOperation as error:
        # Decimal raises InvalidOperation, which nobody expects to catch. Re-raise as the ValueError
        # every caller already handles. `from error` keeps the original in the traceback — dropping
        # it is how you lose the line that actually broke.
        raise ValueError(f"'{text}' is not an amount") from error


def format_money(value: Decimal) -> str:
    """'$1,250.00'. The `,` in the format spec inserts thousands separators; `.2f` fixes it at cents."""
    # A negative reads as -$40.00 rather than $-40.00, which is how a statement is written.
    sign = "-" if value < 0 else ""
    return f"{sign}${abs(value):,.2f}"


def to_csv(value: Decimal) -> str:
    """The plain '1250.00' form written to file — no currency symbol, no separators."""
    return f"{round_money(value):f}"

File: bank/bank-java-console/src/com/bank/model/Money.java
package com.bank.model;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.text.NumberFormat;
import java.util.Locale;

/**
 * Money helpers, kept in one place.
 *
 * <p><b>Never use double for money.</b> `0.1 + 0.2` is `0.30000000000000004` in binary floating
 * point, and a bank that loses a hundredth of a cent per transaction is a bank with a bug. Java's
 * answer is {@link BigDecimal}: exact decimal arithmetic, at the cost of method calls instead of
 * operators (`a.add(b)`, not `a + b`).
 *
 * <p>The class is `final` with a private constructor — the classic way to say "this is a bag of
 * static utilities, do not instantiate or extend it".
 */
public final class Money {

    public static final BigDecimal ZERO = of("0");

    /** Two decimal places, the way currency is written. */
    private static final int SCALE = 2;

    private Money() {
        // Utility class: nothing to construct.
    }

    /**
     * Builds a BigDecimal from a string.
     *
     * <p>Always construct from a String, never `new BigDecimal(0.1)` — the double is already
     * imprecise before BigDecimal ever sees it, so you get 0.1000000000000000055511151231257827.
     */
    public static BigDecimal of(String value) {
        return new BigDecimal(value).setScale(SCALE, RoundingMode.HALF_UP);
    }

    /** Rounds to cents. HALF_UP is what people expect: 2.345 becomes 2.35. */
    public static BigDecimal round(BigDecimal value) {
        return value.setScale(SCALE, RoundingMode.HALF_UP);
    }

    /**
     * Parses user input, tolerating "$1,200.50" as well as "1200.5".
     *
     * @throws NumberFormatException when the text is not a number at all
     */
    public static BigDecimal parse(String input) {
        String cleaned = input.trim().replace("$", "").replace(",", "");
        if (cleaned.isEmpty()) {
            throw new NumberFormatException("No amount entered");
        }
        return round(new BigDecimal(cleaned));
    }

    /** "$1,250.00" — formatting belongs at the edge, next to the screen, not in the model. */
    public static String format(BigDecimal value) {
        return NumberFormat.getCurrencyInstance(Locale.US).format(value);
    }

    /**
     * The plain "1250.00" form written to CSV.
     *
     * <p>toPlainString, not toString: BigDecimal.toString can produce scientific notation
     * ("1.25E+3") for some values, and that is not something the file should ever contain.
     */
    public static String toCsv(BigDecimal value) {
        return round(value).toPlainString();
    }

    public static boolean isPositive(BigDecimal value) {
        // compareTo, never equals: BigDecimal.equals("1.0", "1.00") is false because it compares
        // scale as well as value. This trips up almost everyone once.
        return value.compareTo(BigDecimal.ZERO) > 0;
    }
}
